File: Packages/PassthroughCore/Sources/PassthroughUI/Theme.swift
import SwiftUI

/// The Passthrough visual language: deep graphite surfaces, a teal→violet
/// accent, rounded numerals, and glass cards. Shared by both apps.
public enum PTTheme {
    public static let down = Color(red: 0.25, green: 0.88, blue: 0.82)     // teal: bytes arriving at the Mac
    public static let up = Color(red: 0.71, green: 0.52, blue: 1.0)        // violet: bytes leaving the Mac
    public static let accentStart = down
    public static let accentEnd = up
    public static let warning = Color(red: 1.0, green: 0.72, blue: 0.30)
    public static let danger = Color(red: 1.0, green: 0.42, blue: 0.42)
    public static let success = Color(red: 0.36, green: 0.90, blue: 0.56)

    public static var accent: LinearGradient {
        LinearGradient(colors: [accentStart, accentEnd], startPoint: .topLeading, endPoint: .bottomTrailing)
    }

    public static var accentAngular: AngularGradient {
        AngularGradient(colors: [accentStart, accentEnd, accentStart], center: .center)
    }

    public static func display(_ size: CGFloat, weight: Font.Weight = .bold) -> Font {
        .system(size: size, weight: weight, design: .rounded)
    }

    public static func mono(_ size: CGFloat, weight: Font.Weight = .medium) -> Font {
        .system(size: size, weight: weight, design: .rounded).monospacedDigit()
    }
}

/// Background: near-black graphite with two soft accent glows, adapting to light mode.
public struct PTBackground: View {
    @Environment(\.colorScheme) private var scheme
    public var glow: Double
    public init(glow: Double = 1) { self.glow = glow }

    public var body: some View {
        ZStack {
            (scheme == .dark ? Color(red: 0.05, green: 0.06, blue: 0.09) : Color(red: 0.95, green: 0.96, blue: 0.98))
            GeometryReader { geo in
                Circle()
                    .fill(PTTheme.accentStart.opacity((scheme == .dark ? 0.22 : 0.18) * glow))
                    .frame(width: geo.size.width * 0.9)
                    .blur(radius: 90)
                    .offset(x: -geo.size.width * 0.35, y: -geo.size.height * 0.25)
                Circle()
                    .fill(PTTheme.accentEnd.opacity((scheme == .dark ? 0.20 : 0.16) * glow))
                    .frame(width: geo.size.width * 0.8)
                    .blur(radius: 100)
                    .offset(x: geo.size.width * 0.45, y: geo.size.height * 0.45)
            }
            // Rasterise the two blurred glows on the GPU once per change instead
            // of convolving them on the CPU every time the view updates.
            .drawingGroup()
        }
        .ignoresSafeArea()
        .animation(.easeInOut(duration: 1.2), value: glow)
    }
}

/// Glass card container.
public struct PTCard<Content: View>: View {
    @Environment(\.colorScheme) private var scheme
    private let content: Content
    private let padding: CGFloat
    public init(padding: CGFloat = 16, @ViewBuilder content: () -> Content) {
        self.content = content()
        self.padding = padding
    }
    public var body: some View {
        content
            .padding(padding)
            .background {
                RoundedRectangle(cornerRadius: 22, style: .continuous)
                    .fill(.regularMaterial)
                    .overlay(
                        RoundedRectangle(cornerRadius: 22, style: .continuous)
                            .strokeBorder(Color.white.opacity(scheme == .dark ? 0.08 : 0.5), lineWidth: 1)
                    )
                    .shadow(color: .black.opacity(scheme == .dark ? 0.35 : 0.08), radius: 18, y: 8)
            }
    }
}

public struct PTSectionTitle: View {
    let text: String
    let icon: String?
    public init(_ text: String, icon: String? = nil) { self.text = text; self.icon = icon }
    public var body: some View {
        HStack(spacing: 6) {
            if let icon { Image(systemName: icon).font(.caption.weight(.semibold)) }
            Text(text.uppercased()).font(.caption.weight(.semibold)).tracking(1.1)
        }
        .foregroundStyle(.secondary)
    }
}

/// Small capsule chip, e.g. "5G", "USB", "VPN".
public struct PTPill: View {
    let text: String
    let icon: String?
    let tint: Color
    public init(_ text: String, icon: String? = nil, tint: Color = .secondary) { self.text = text; self.icon = icon; self.tint = tint }
    public var body: some View {
        HStack(spacing: 4) {
            if let icon { Image(systemName: icon).font(.caption2.weight(.bold)) }
            Text(text).font(.caption.weight(.semibold))
        }
        .padding(.horizontal, 9).padding(.vertical, 5)
        .foregroundStyle(tint)
        .background(tint.opacity(0.14), in: Capsule())
    }
}

/// Status dot with a soft halo, pulsing when live.
public struct PTStatusDot: View {
    let color: Color
    let live: Bool
    @State private var pulse = false
    public init(color: Color, live: Bool) { self.color = color; self.live = live }
    public var body: some View {
        ZStack {
            Circle().fill(color.opacity(0.35)).frame(width: 16, height: 16)
                .scaleEffect(live && pulse ? 1.5 : 1).opacity(live && pulse ? 0 : 1)
            Circle().fill(color).frame(width: 8, height: 8)
        }
        .onAppear { withAnimation(.easeOut(duration: 1.4).repeatForever(autoreverses: false)) { pulse = true } }
    }
}

File: android/app/src/main/java/dev/dpatel/passthrough/ui/Theme.kt
package dev.dpatel.passthrough.ui

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.getValue
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.ExperimentalTextApi
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontVariation
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import dev.dpatel.passthrough.R

/**
 * The Passthrough visual language, shared with the iOS and Mac apps: deep
 * graphite surfaces, a teal→violet accent, rounded numerals and glass cards.
 */
object PT {
    val down = Color(0xFF40E0D1)      // teal: bytes arriving at the Mac
    val up = Color(0xFFB585FF)        // violet: bytes leaving the Mac
    val accentStart = down
    val accentEnd = up
    val warning = Color(0xFFFFB84D)
    val danger = Color(0xFFFF6B6B)
    val success = Color(0xFF5CE68F)

    val accent: Brush get() = Brush.linearGradient(listOf(accentStart, accentEnd))

    @OptIn(ExperimentalTextApi::class)
    private fun nunito(weight: FontWeight) = Font(
        R.font.nunito, weight = weight,
        variationSettings = FontVariation.Settings(FontVariation.weight(weight.weight)),
    )

    val rounded = FontFamily(
        nunito(FontWeight.Normal), nunito(FontWeight.Medium), nunito(FontWeight.SemiBold),
        nunito(FontWeight.Bold), nunito(FontWeight.ExtraBold),
    )

    fun display(size: TextUnit, weight: FontWeight = FontWeight.Bold) = TextStyle(fontFamily = rounded, fontSize = size, fontWeight = weight)

    /** Rounded with tabular figures, so changing numbers don't jiggle. */
    fun mono(size: TextUnit, weight: FontWeight = FontWeight.Medium) =
        TextStyle(fontFamily = rounded, fontSize = size, fontWeight = weight, fontFeatureSettings = "tnum")
}

/** Label colours matching iOS's primary/secondary/tertiary for the current scheme. */
data class PTColors(val dark: Boolean) {
    val primary = if (dark) Color.White else Color.Black
    val secondary = if (dark) Color(0x99EBEBF5) else Color(0x993C3C43)
    val tertiary = if (dark) Color(0x4DEBEBF5) else Color(0x4D3C3C43)
    val background = if (dark) Color(0xFF0D0F17) else Color(0xFFF2F5FA)
    val card = if (dark) Color(0xB31C1F2B) else Color(0xCCFFFFFF)
    val cardBorder = if (dark) Color.White.copy(alpha = 0.08f) else Color.White.copy(alpha = 0.5f)
    val fill = if (dark) Color.White.copy(alpha = 0.08f) else Color.Black.copy(alpha = 0.06f)
    val thinFill = if (dark) Color.White.copy(alpha = 0.06f) else Color.White.copy(alpha = 0.7f)
}

val LocalPT = staticCompositionLocalOf { PTColors(true) }

@Composable
fun PassthroughTheme(content: @Composable () -> Unit) {
    val dark = isSystemInDarkTheme()
    val colors = PTColors(dark)
    val scheme = if (dark) darkColorScheme(
        primary = PT.accentEnd, secondary = PT.accentStart, background = colors.background, surface = Color(0xFF161925),
        onPrimary = Color.White, onBackground = Color.White, onSurface = Color.White, error = PT.danger,
        surfaceContainerHigh = Color(0xFF1E2130), surfaceContainerLow = Color(0xFF161925),
    ) else lightColorScheme(
        primary = Color(0xFF7B4FE0), secondary = Color(0xFF14A89B), background = colors.background, surface = Color.White,
        onPrimary = Color.White, onBackground = Color.Black, onSurface = Color.Black, error = Color(0xFFD93A3A),
    )
    val base = MaterialTheme.typography
    val typography = base.copy(
        headlineLarge = base.headlineLarge.copy(fontFamily = PT.rounded),
        titleLarge = base.titleLarge.copy(fontFamily = PT.rounded, fontWeight = FontWeight.Bold),
        titleMedium = base.titleMedium.copy(fontFamily = PT.rounded, fontWeight = FontWeight.SemiBold),
        bodyLarge = base.bodyLarge.copy(fontFamily = PT.rounded),
        bodyMedium = base.bodyMedium.copy(fontFamily = PT.rounded),
        bodySmall = base.bodySmall.copy(fontFamily = PT.rounded),
        labelLarge = base.labelLarge.copy(fontFamily = PT.rounded, fontWeight = FontWeight.SemiBold),
        labelMedium = base.labelMedium.copy(fontFamily = PT.rounded),
        labelSmall = base.labelSmall.copy(fontFamily = PT.rounded),
    )
    CompositionLocalProvider(LocalPT provides colors) {
        MaterialTheme(colorScheme = scheme, typography = typography, content = content)
    }
}

/** Near-black graphite (or pale grey) with two soft accent glows. */
@Composable
fun PTBackground(glow: Float, modifier: Modifier = Modifier, content: @Composable BoxScope.() -> Unit = {}) {
    val c = LocalPT.current
    val g by animateFloatAsState(glow, tween(1200), label = "glow")
    Box(
        modifier
            .fillMaxSize()
            .background(c.background)
            .drawBehind {
                val w = size.width
                val h = size.height
                val a1 = (if (c.dark) 0.22f else 0.18f) * g
                val a2 = (if (c.dark) 0.20f else 0.16f) * g
                drawCircle(
                    Brush.radialGradient(listOf(PT.accentStart.copy(alpha = a1), Color.Transparent), center = Offset(w * 0.15f, h * 0.25f), radius = w * 0.85f),
                    radius = w * 0.85f, center = Offset(w * 0.15f, h * 0.25f),
                )
                drawCircle(
                    Brush.radialGradient(listOf(PT.accentEnd.copy(alpha = a2), Color.Transparent), center = Offset(w * 0.95f, h * 0.95f), radius = w * 0.8f),
                    radius = w * 0.8f, center = Offset(w * 0.95f, h * 0.95f),
                )
            },
        content = content,
    )
}

/** Glass card container. */
@Composable
fun PTCard(modifier: Modifier = Modifier, padding: Dp = 16.dp, content: @Composable ColumnScope.() -> Unit) {
    val c = LocalPT.current
    val shape = RoundedCornerShape(22.dp)
    Column(
        modifier
            .fillMaxWidth()
            .then(if (c.dark) Modifier else Modifier.shadow(14.dp, shape, ambientColor = Color.Black.copy(alpha = 0.08f), spotColor = Color.Black.copy(alpha = 0.08f)))
            .background(c.card, shape)
            .border(1.dp, c.cardBorder, shape)
            .padding(PaddingValues(padding)),
        content = content,
    )
}

